language/move-symbol-pool/src/pool.rs

```diff
@@ -0,0 +1,120 @@
+//!   a pointer would not be safe if the data being pointed to could be
+//!   reallocated.
+//! * Why not use a [`LinkedList`]? A linked list does not unique the elements.
+//!   Ensuring the elements in the list are unique would require traversing the
+//!   list, which would not be performant for large lists.
+//! * Why not use a [`HashSet`] in conjunction with a [`LinkedList`]? This would
+//!   be simpler to implement, but would involve storing the string data twice:
+//!   once in the set for uniqueness checking, and once in the linked list to
+//!   maintain a constant memory address for the string data. This pool
+//!   implementation is more space-efficient.
+
+use std::{
+    collections::hash_map::DefaultHasher,
+    hash::{Hash, Hasher},
+    ptr::NonNull,
+};
+
+/// The number of buckets in the pool's contiguous array.
+const NB_BUCKETS: usize = 1 << 12; // 4096
+/// A mask used to convert a string hash into the index of one of the contiguous buckets.
+const BUCKET_MASK: u64 = NB_BUCKETS as u64 - 1;
+
+/// A bucket is a space on the heap within which an entry may be allocated.
+type Bucket = Option<Box<Entry>>;
+
+/// A string in the pool.
+pub(crate) struct Entry {
+    pub(crate) string: Box<str>,
+    hash: u64,
+    next: Bucket,
+}
+
+/// A contiguous array of buckets.
+pub(crate) struct Pool(Box<[Bucket; NB_BUCKETS]>);
+
+impl Pool {
+    /// Allocates a contiguous array of buckets on the heap. As strings are
+    /// inserted into the pool, buckets in this array are filled with an entry.
+    pub(crate) fn new() -> Self {
+        let vec = std::mem::ManuallyDrop::new(vec![0_usize; NB_BUCKETS]);
+        Self(unsafe { Box::from_raw(vec.as_ptr() as *mut [Bucket; NB_BUCKETS]) })
+    }
+
+    /// Computes the hash value of a string, which is used to determine both
+    /// which top-level bucket contains an entry corresponding to the string,
+    /// as well as a scalar value that can be used to quickly check whether
+    /// two strings are not equal.
+    fn hash(string: &str) -> u64 {
+        let mut hasher = DefaultHasher::new();
+        string.hash(&mut hasher);
+        hasher.finish()
+    }
+
+    /// Given a string, returns its entry in the pool (adding it if it does not
+    /// yet exist in the pool).
+    pub(crate) fn insert(&mut self, string: &str) -> NonNull<Entry> {
+        let hash = Self::hash(string);
+        // Access the top-level bucket in the pool's contiguous array that
+        // contains the linked list of entries that contain the string.
+        let bucket_index = (hash & BUCKET_MASK) as usize;
+        let mut ptr: Option<&mut Box<Entry>> = self.0[bucket_index].as_mut();
+
+        // Iterate over the entires in the bucket.
+        while let Some(entry) = ptr.take() {
+            // If we find the string we're looking for, don't add anything to
+            // the pool. Instead, just return the existing entry.
+            // NOTE: Strings that are not equal must have different hash values,
```

Comment:
I think this needs to be reworded as, "different hash values always mean different strings"
Since Strings that are not equal _might_ have the same hash (hence the second check)